File: harness/decision_capsule_contract/manifest.py
"""StructuralReplayManifest derivation, sealing, and validation."""

from __future__ import annotations

import hashlib

from kernel_decision_contract import validate_closure as validate_decision_closure
from kernel_decision_contract.constants import (
    MAX_CLOSURE_BYTES as DECISION_CLOSURE_MAX_BYTES,
)
from kernel_operational_contract.shape import bare_hash

from .codec import (ContractError, bounded_deepcopy, canonical_json,
                    decode_canonical_json, exact_object_shell,
                    require_constants, require_identity_strings,
                    validate_json_tree)
from .constants import *  # noqa: F403 - frozen vocabulary is intentionally central
from .shape import (artifact_receipt_refs, artifact_refs, atom_refs, attestations,
                    decision_closure_ref, event_refs, execution_refs,
                    invocation_refs, operational_closure_ref, transaction_ref)


def _record_ref(record: dict, id_field: str, hash_field: str) -> dict[str, object]:
    return {id_field: record[id_field], hash_field: record[hash_field]}


def _candidate_validated(closure: dict[str, object], *, copy_artifacts: bool = False
                         ) -> dict[str, object]:
    transaction, operational = closure["decision_transaction"], closure["operational_closure"]
    atoms = closure["cognitive_atoms"]
    artifacts = operational["artifacts"]
    if copy_artifacts:
        artifacts = bounded_deepcopy(artifacts, MAX_MANIFEST_BYTES)  # noqa: F405
    return {
        "api_version": MANIFEST_API,  # noqa: F405
        "artifact_receipt_refs": [_record_ref(item, "artifact_receipt_id",
                                              "artifact_receipt_sha256")
                                  for item in operational["artifact_receipts"]],
        "artifact_refs": artifacts,
        "attestations": {field: False for field in ATTESTATION_FIELDS},  # noqa: F405
        "canonicalization": CANONICALIZATION,  # noqa: F405
        "capability_invocation_refs": [_record_ref(item, "invocation_id",
                                                   "invocation_sha256")
                                       for item in operational["capability_invocations"]],
        "decision_closure_ref": _record_ref(closure, "closure_id", "closure_sha256"),
        "decision_transaction_ref": _record_ref(
            transaction, "decision_transaction_id", "decision_transaction_sha256"),
        "effect_replay_allowed": False,
        "execution_receipt_refs": [_record_ref(item, "execution_receipt_id",
                                               "execution_receipt_sha256")
                                   for item in operational["execution_receipts"]],
        "history_rewrite_allowed": False,
        "interaction_event_refs": [_record_ref(item, "event_id", "event_sha256")
                                   for item in operational["interaction_events"]],
        "kind": MANIFEST_KIND, "manifest_id": "", "manifest_sha256": "",  # noqa: F405
        "operational_closure_ref": _record_ref(
            operational, "closure_id", "closure_sha256"),
        "postdecision_atom_refs": [_record_ref(item, "atom_id", "atom_sha256")
                                   for item in atoms
                                   if item["source"]["source_phase"] == "postdecision"],
        "predecision_atom_refs": [_record_ref(item, "atom_id", "atom_sha256")
                                  for item in atoms
                                  if item["source"]["source_phase"] == "predecision"],
        "replay_mode": MANIFEST_MODE,  # noqa: F405
    }


def _identity(manifest: dict[str, object], allow_blank: bool) -> None:
    if allow_blank and manifest["manifest_id"] == manifest["manifest_sha256"] == "":
        return
    digest = bare_hash(manifest["manifest_sha256"], "manifest_sha256")
    if manifest["manifest_id"] != f"{MANIFEST_PREFIX}{digest}":  # noqa: F405
        raise ContractError("manifest_id must bind manifest_sha256")


def _local_preflight(value: object, *, allow_blank: bool = False,
                     ignore_identity: bool = False) -> dict[str, object]:
    manifest = exact_object_shell(  # noqa: F405
        value, MANIFEST_FIELDS, "StructuralReplayManifest")
    constants = {"api_version": MANIFEST_API, "canonicalization": CANONICALIZATION,  # noqa: F405
                 "kind": MANIFEST_KIND, "replay_mode": MANIFEST_MODE}  # noqa: F405
    require_constants(manifest, constants)
    if manifest["effect_replay_allowed"] is not False:
        raise ContractError("effect_replay_allowed must be false")
    if manifest["history_rewrite_allowed"] is not False:
        raise ContractError("history_rewrite_allowed must be false")
    require_identity_strings(manifest, "manifest_id", "manifest_sha256")
    validate_json_tree(manifest, MAX_MANIFEST_BYTES)  # noqa: F405
    if not ignore_identity:
        _identity(manifest, allow_blank)
    attestations(manifest["attestations"])
    decision_closure_ref(manifest["decision_closure_ref"])
    transaction_ref(manifest["decision_transaction_ref"])
    operational_closure_ref(manifest["operational_closure_ref"])
    predecision = atom_refs(
        manifest["predecision_atom_refs"], "predecision_atom_refs", 1)
    postdecision = atom_refs(
        manifest["postdecision_atom_refs"], "postdecision_atom_refs", 0)
    if len(predecision) + len(postdecision) > MAX_ATOMS:  # noqa: F405
        raise ContractError("combined decision atom refs exceed 256")
    atom_ids = [item["atom_id"] for item in predecision + postdecision]
    if len(atom_ids) != len(set(atom_ids)):
        raise ContractError("combined decision atom refs must be unique")
    artifact_refs(manifest["artifact_refs"], "artifact_refs", MAX_ARTIFACTS)  # noqa: F405
    artifact_receipt_refs(manifest["artifact_receipt_refs"])
    invocation_refs(manifest["capability_invocation_refs"])
    event_refs(manifest["interaction_event_refs"])
    execution_refs(manifest["execution_receipt_refs"])
    return manifest


def _shape(value: object, *, allow_blank: bool = False) -> dict[str, object]:
    return _local_preflight(value, allow_blank=allow_blank)


def _blanked(value: object) -> dict[str, object]:
    exact_object_shell(value, MANIFEST_FIELDS, "manifest digest input")  # noqa: F405
    validate_json_tree(value, MAX_MANIFEST_BYTES)  # noqa: F405
    result = bounded_deepcopy(value, MAX_MANIFEST_BYTES)  # noqa: F405
    result["manifest_id"], result["manifest_sha256"] = "", ""
    return result


def _validate_projection(manifest: dict[str, object], closure: dict[str, object]) -> None:
    try:
        expected = _candidate_validated(closure)
    except (KeyError, TypeError) as error:
        raise ContractError(
            "decision closure cannot supply the manifest projection") from error
    identity = {"manifest_id", "manifest_sha256"}
    if any(manifest[field] != expected[field] for field in MANIFEST_FIELDS - identity):  # noqa: F405
        raise ContractError("manifest must be the exact ordered projection of its closure")


def _validated_dependency(manifest: dict[str, object],
                          decision_closure: object) -> dict[str, object]:
    validate_json_tree(decision_closure, DECISION_CLOSURE_MAX_BYTES)
    _validate_projection(manifest, decision_closure)
    closure = validate_decision_closure(decision_closure)
    _validate_projection(manifest, closure)
    return closure


def _digest_validated(blank: dict[str, object]) -> str:
    raw = canonical_json(blank)
    if len(raw) > MAX_MANIFEST_BYTES:  # noqa: F405
        raise ContractError("manifest blank preimage exceeds its byte ceiling")
    return hashlib.sha256(MANIFEST_DOMAIN + raw).hexdigest()  # noqa: F405


def _validate_with_validated_closure(
        value: object, closure: dict[str, object], *, allow_blank: bool = False
        ) -> dict[str, object]:
    manifest = _shape(value, allow_blank=allow_blank)
    return _validate_shaped_with_validated_closure(
        manifest, closure, allow_blank=allow_blank)


def _validate_shaped_with_validated_closure(
        manifest: dict[str, object], closure: dict[str, object], *,
        allow_blank: bool = False) -> dict[str, object]:
    _validate_projection(manifest, closure)
    blank = _blanked(manifest)
    digest = _digest_validated(blank)
    if len(canonical_json(manifest)) > MAX_MANIFEST_BYTES:  # noqa: F405
        raise ContractError("manifest exceeds its sealed byte ceiling")
    if not allow_blank and manifest["manifest_sha256"] != digest:
        raise ContractError("manifest_sha256 does not match canonical preimage")
    return manifest


def _derive_with_validated_closure(closure: dict[str, object]) -> dict[str, object]:
    manifest = _candidate_validated(closure, copy_artifacts=True)
    _shape(manifest, allow_blank=True)
    digest = _digest_validated(manifest)
    manifest["manifest_id"] = f"{MANIFEST_PREFIX}{digest}"  # noqa: F405
    manifest["manifest_sha256"] = digest
    if len(canonical_json(manifest)) > MAX_MANIFEST_BYTES:  # noqa: F405
        raise ContractError("manifest exceeds its sealed byte ceiling")
    _identity(manifest, False)
    return manifest


def structural_replay_manifest_digest(value: object,
                                      decision_closure: object) -> str:
    manifest = _local_preflight(value, ignore_identity=True)
    _validated_dependency(manifest, decision_closure)
    blank = _blanked(manifest)
    _shape(blank, allow_blank=True)
    return _digest_validated(blank)


def validate_structural_replay_manifest(value: object,
                                        decision_closure: object) -> dict[str, object]:
    manifest = _shape(value)
    closure = _validated_dependency(manifest, decision_closure)
    return _validate_shaped_with_validated_closure(manifest, closure)


def seal_structural_replay_manifest(value: object,
                                    decision_closure: object) -> dict[str, object]:
    if (type(value) is not dict or value.get("manifest_id") != "" or
            value.get("manifest_sha256") != ""):
        raise ContractError("sealing manifest requires blank own identity")
    shaped = _shape(value, allow_blank=True)
    closure = _validated_dependency(shaped, decision_closure)
    _validate_shaped_with_validated_closure(shaped, closure, allow_blank=True)
    manifest = bounded_deepcopy(shaped, MAX_MANIFEST_BYTES)  # noqa: F405
    digest = _digest_validated(manifest)
    manifest["manifest_id"] = f"{MANIFEST_PREFIX}{digest}"  # noqa: F405
    manifest["manifest_sha256"] = digest
    if len(canonical_json(manifest)) > MAX_MANIFEST_BYTES:  # noqa: F405
        raise ContractError("manifest exceeds its sealed byte ceiling")
    _identity(manifest, False)
    return manifest


def derive_structural_replay_manifest(decision_closure: object) -> dict[str, object]:
    """Derive the only manifest admitted for one sealed ADR-0090 closure."""
    validate_json_tree(decision_closure, DECISION_CLOSURE_MAX_BYTES)
    closure = validate_decision_closure(decision_closure)
    return _derive_with_validated_closure(closure)


def decode_structural_replay_manifest(raw: bytes,
                                      decision_closure: object) -> dict[str, object]:
    value = decode_canonical_json(raw, MAX_MANIFEST_BYTES)  # noqa: F405
    return validate_structural_replay_manifest(value, decision_closure)

File: forge-runtime/crates/domain/src/decision_capsule_contract/manifest.rs
use crate::kernel_decision_contract::{AtomRef, KernelDecisionReferenceClosure};
use std::collections::HashSet;

use super::{
    CANONICALIZATION, DecisionCapsuleContractError, MANIFEST_API, MANIFEST_DOMAIN, MANIFEST_KIND,
    MANIFEST_MODE, MANIFEST_PREFIX, MAX_MANIFEST_BYTES, StructuralReplayManifest, invalid,
    primitives, wire,
};

fn header_projection_matches(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
) -> bool {
    let operational = &closure.operational_closure;
    let transaction = &closure.decision_transaction;
    value.decision_closure_ref.closure_id == closure.closure_id
        && value.decision_closure_ref.closure_sha256 == closure.closure_sha256
        && value.decision_transaction_ref.decision_transaction_id
            == transaction.decision_transaction_id
        && value.decision_transaction_ref.decision_transaction_sha256
            == transaction.decision_transaction_sha256
        && value.operational_closure_ref.closure_id == operational.closure_id
        && value.operational_closure_ref.closure_sha256 == operational.closure_sha256
}

fn operational_projection_matches(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
) -> bool {
    let operational = &closure.operational_closure;
    let receipts = value
        .artifact_receipt_refs
        .iter()
        .map(|item| (&item.artifact_receipt_id, &item.artifact_receipt_sha256));
    let expected_receipts = operational
        .artifact_receipts
        .iter()
        .map(|item| (&item.artifact_receipt_id, &item.artifact_receipt_sha256));
    let invocations = value
        .capability_invocation_refs
        .iter()
        .map(|item| (&item.invocation_id, &item.invocation_sha256));
    let expected_invocations = operational
        .capability_invocations
        .iter()
        .map(|item| (&item.invocation_id, &item.invocation_sha256));
    value.artifact_refs == operational.artifacts
        && receipts.eq(expected_receipts)
        && invocations.eq(expected_invocations)
        && event_receipt_projection_matches(value, closure)
}

fn event_receipt_projection_matches(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
) -> bool {
    let operational = &closure.operational_closure;
    let events = value
        .interaction_event_refs
        .iter()
        .map(|item| (&item.event_id, &item.event_sha256));
    let expected_events = operational
        .interaction_events
        .iter()
        .map(|item| (&item.event_id, &item.event_sha256));
    let executions = value
        .execution_receipt_refs
        .iter()
        .map(|item| (&item.execution_receipt_id, &item.execution_receipt_sha256));
    let expected_executions = operational
        .execution_receipts
        .iter()
        .map(|item| (&item.execution_receipt_id, &item.execution_receipt_sha256));
    events.eq(expected_events) && executions.eq(expected_executions)
}

fn atom_projection_matches(
    actual: &[AtomRef],
    closure: &KernelDecisionReferenceClosure,
    phase: &str,
) -> bool {
    let actual = actual.iter().map(|item| (&item.atom_id, &item.atom_sha256));
    let expected = closure
        .cognitive_atoms
        .iter()
        .filter(|atom| atom.source.source_phase == phase)
        .map(|item| (&item.atom_id, &item.atom_sha256));
    actual.eq(expected)
}

fn projection_matches(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
) -> bool {
    header_projection_matches(value, closure)
        && operational_projection_matches(value, closure)
        && atom_projection_matches(&value.predecision_atom_refs, closure, "predecision")
        && atom_projection_matches(&value.postdecision_atom_refs, closure, "postdecision")
}

pub(super) fn require_projection(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
) -> Result<(), DecisionCapsuleContractError> {
    if projection_matches(value, closure) {
        Ok(())
    } else {
        Err(invalid(
            "manifest is not the exact ordered closure projection",
        ))
    }
}

fn references<'a>(
    values: impl IntoIterator<Item = (&'a str, &'a str)>,
    prefix: &str,
    label: &str,
    sorted: bool,
) -> Result<(), DecisionCapsuleContractError> {
    let mut identities = Vec::new();
    for (identity, digest) in values {
        primitives::identity(identity, digest, prefix, label, false)?;
        identities.push(identity);
    }
    let mut seen = HashSet::with_capacity(identities.len());
    if identities.iter().any(|identity| !seen.insert(*identity)) {
        return Err(invalid(format!("{label} must be unique")));
    }
    if sorted
        && !identities
            .windows(2)
            .all(|pair| pair[0].as_bytes() < pair[1].as_bytes())
    {
        return Err(invalid(format!("{label} must preserve identity order")));
    }
    Ok(())
}

fn validate_header_refs(
    value: &StructuralReplayManifest,
) -> Result<(), DecisionCapsuleContractError> {
    primitives::identity(
        &value.decision_closure_ref.closure_id,
        &value.decision_closure_ref.closure_sha256,
        "kernel-decision-reference-closure-",
        "decision_closure_ref",
        false,
    )?;
    primitives::identity(
        &value.decision_transaction_ref.decision_transaction_id,
        &value.decision_transaction_ref.decision_transaction_sha256,
        "decision-transaction-",
        "decision_transaction_ref",
        false,
    )?;
    primitives::identity(
        &value.operational_closure_ref.closure_id,
        &value.operational_closure_ref.closure_sha256,
        "kernel-operational-reference-closure-",
        "operational_closure_ref",
        false,
    )
}

fn validate_atom_refs(
    value: &StructuralReplayManifest,
) -> Result<(), DecisionCapsuleContractError> {
    let pre = value
        .predecision_atom_refs
        .iter()
        .map(|item| (item.atom_id.as_str(), item.atom_sha256.as_str()));
    references(pre, "cognitive-atom-", "predecision_atom_refs", true)?;
    let post = value
        .postdecision_atom_refs
        .iter()
        .map(|item| (item.atom_id.as_str(), item.atom_sha256.as_str()));
    references(post, "cognitive-atom-", "postdecision_atom_refs", true)?;
    let combined = value
        .predecision_atom_refs
        .iter()
        .chain(&value.postdecision_atom_refs)
        .map(|item| (item.atom_id.as_str(), item.atom_sha256.as_str()));
    references(combined, "cognitive-atom-", "combined atom refs", false)
}

fn validate_artifact_refs(
    value: &StructuralReplayManifest,
) -> Result<(), DecisionCapsuleContractError> {
    let mut encoded = Vec::with_capacity(value.artifact_refs.len());
    for artifact in &value.artifact_refs {
        primitives::validate_artifact(artifact)?;
        encoded.push(wire::canonical_with_max(artifact, MAX_MANIFEST_BYTES)?);
    }
    if encoded.windows(2).all(|pair| pair[0] < pair[1]) {
        Ok(())
    } else {
        Err(invalid(
            "artifact_refs must be canonical-byte sorted and unique",
        ))
    }
}

fn validate_attempt_refs(
    value: &StructuralReplayManifest,
) -> Result<(), DecisionCapsuleContractError> {
    let receipts = value.artifact_receipt_refs.iter().map(|item| {
        (
            item.artifact_receipt_id.as_str(),
            item.artifact_receipt_sha256.as_str(),
        )
    });
    references(receipts, "artifact-receipt-", "artifact_receipt_refs", true)?;
    let invocations = value
        .capability_invocation_refs
        .iter()
        .map(|item| (item.invocation_id.as_str(), item.invocation_sha256.as_str()));
    references(
        invocations,
        "capability-invocation-",
        "capability_invocation_refs",
        false,
    )
}

fn validate_event_receipt_refs(
    value: &StructuralReplayManifest,
) -> Result<(), DecisionCapsuleContractError> {
    let events = value
        .interaction_event_refs
        .iter()
        .map(|item| (item.event_id.as_str(), item.event_sha256.as_str()));
    references(
        events,
        "interaction-event-",
        "interaction_event_refs",
        false,
    )?;
    let executions = value.execution_receipt_refs.iter().map(|item| {
        (
            item.execution_receipt_id.as_str(),
            item.execution_receipt_sha256.as_str(),
        )
    });
    references(
        executions,
        "execution-receipt-",
        "execution_receipt_refs",
        false,
    )
}

fn body_shape(value: &StructuralReplayManifest) -> Result<(), DecisionCapsuleContractError> {
    // Digest ignores the values of the two own identity fields, but their JSON
    // profile limits still precede local canonical-key work and all dependency
    // validation.
    wire::premeasure_only(
        &(&value.manifest_id, &value.manifest_sha256),
        MAX_MANIFEST_BYTES,
    )?;
    if value.api_version != MANIFEST_API
        || value.canonicalization != CANONICALIZATION
        || value.kind != MANIFEST_KIND
        || value.replay_mode != MANIFEST_MODE
    {
        return Err(invalid("StructuralReplayManifest constants differ"));
    }
    if value.effect_replay_allowed || value.history_rewrite_allowed {
        return Err(invalid("manifest replay controls must be false"));
    }
    primitives::attestations(&value.attestations)?;
    let atoms = value.predecision_atom_refs.len() + value.postdecision_atom_refs.len();
    if value.predecision_atom_refs.is_empty()
        || atoms > 256
        || value.artifact_refs.len() > 256
        || value.artifact_receipt_refs.len() > 64
        || !(1..=64).contains(&value.capability_invocation_refs.len())
        || value.interaction_event_refs.len() > 256
        || !(1..=64).contains(&value.execution_receipt_refs.len())
    {
        return Err(invalid("manifest inventory cardinality differs"));
    }
    validate_header_refs(value)?;
    validate_atom_refs(value)?;
    validate_artifact_refs(value)?;
    validate_attempt_refs(value)?;
    validate_event_receipt_refs(value)
}

fn shape(
    value: &StructuralReplayManifest,
    allow_blank: bool,
) -> Result<(), DecisionCapsuleContractError> {
    body_shape(value)?;
    primitives::identity(
        &value.manifest_id,
        &value.manifest_sha256,
        MANIFEST_PREFIX,
        "manifest",
        allow_blank,
    )
}

pub(super) fn local_shape(
    value: &StructuralReplayManifest,
    allow_blank: bool,
) -> Result<(), DecisionCapsuleContractError> {
    shape(value, allow_blank)
}

fn blanked_after_validation(
    value: &StructuralReplayManifest,
) -> Result<StructuralReplayManifest, DecisionCapsuleContractError> {
    wire::premeasure_only(value, MAX_MANIFEST_BYTES)?;
    let mut blank = wire::bounded_clone(value, MAX_MANIFEST_BYTES)?;
    blank.manifest_id.clear();
    blank.manifest_sha256.clear();
    Ok(blank)
}

pub(super) fn validate_with_closure(
    value: &StructuralReplayManifest,
    closure: &KernelDecisionReferenceClosure,
    allow_blank: bool,
) -> Result<String, DecisionCapsuleContractError> {
    body_shape(value)?;
    wire::premeasure_only(value, MAX_MANIFEST_BYTES)?;
    if !allow_blank {
        primitives::identity(
            &value.manifest_id,
            &value.manifest_sha256,
            MANIFEST_PREFIX,
            "manifest",
            false,
        )?;
    }
    require_projection(value, closure)?;
    let blank = blanked_after_validation(value)?;
    let checked = if allow_blank { &blank } else { value };
    let digest = wire::domain_digest(MANIFEST_DOMAIN, &blank, MAX_MANIFEST_BYTES)?;
    wire::canonical_with_max(checked, MAX_MANIFEST_BYTES)?;
    if !allow_blank && value.manifest_sha256 != digest {
        return Err(invalid("manifest_sha256 differs from canonical preimage"));
    }
    Ok(digest)
}

pub(super) fn derive_with_closure(
    closure: &KernelDecisionReferenceClosure,
) -> Result<StructuralReplayManifest, DecisionCapsuleContractError> {
    let mut manifest = primitives::project_manifest(closure);
    let digest = wire::domain_digest(MANIFEST_DOMAIN, &manifest, MAX_MANIFEST_BYTES)?;
    manifest.manifest_id = format!("{MANIFEST_PREFIX}{digest}");
    manifest.manifest_sha256 = digest;
    wire::canonical_with_max(&manifest, MAX_MANIFEST_BYTES)?;
    shape(&manifest, false)?;
    Ok(manifest)
}

/// Computes the manifest digest after validating its sealed ADR-0090 dependency.
///
/// # Errors
///
/// Returns an error for an invalid dependency, projection, identity, or document ceiling.
pub fn structural_replay_manifest_digest(
    value: &StructuralReplayManifest,
    decision_closure: &KernelDecisionReferenceClosure,
) -> Result<String, DecisionCapsuleContractError> {
    body_shape(value)?;
    wire::premeasure_only(value, MAX_MANIFEST_BYTES)?;
    require_projection(value, decision_closure)?;
    let closure = primitives::reseal_decision_closure(decision_closure)?;
    validate_with_closure(value, &closure, true)
}

/// Validates the exact manifest projection and its sealed ADR-0090 dependency.
///
/// # Errors
///
/// Returns an error for an invalid dependency, projection, identity, or document ceiling.
pub fn validate_structural_replay_manifest(
    value: &StructuralReplayManifest,
    decision_closure: &KernelDecisionReferenceClosure,
) -> Result<(), DecisionCapsuleContractError> {
    shape(value, false)?;
    wire::premeasure_only(value, MAX_MANIFEST_BYTES)?;
    require_projection(value, decision_closure)?;
    let closure = primitives::reseal_decision_closure(decision_closure)?;
    validate_with_closure(value, &closure, false).map(|_| ())
}

/// Seals a blank manifest only when it exactly projects the supplied ADR-0090 closure.
///
/// # Errors
///
/// Returns an error for a nonblank identity, invalid dependency or projection, or oversize.
pub fn seal_structural_replay_manifest(
    value: &StructuralReplayManifest,
    decision_closure: &KernelDecisionReferenceClosure,
) -> Result<StructuralReplayManifest, DecisionCapsuleContractError> {
    if !value.manifest_id.is_empty() || !value.manifest_sha256.is_empty() {
        return Err(invalid("sealing manifest requires blank own identity"));
    }
    shape(value, true)?;
    wire::premeasure_only(value, MAX_MANIFEST_BYTES)?;
    require_projection(value, decision_closure)?;
    let closure = primitives::reseal_decision_closure(decision_closure)?;
    let digest = validate_with_closure(value, &closure, true)?;
    let mut sealed = wire::bounded_clone(value, MAX_MANIFEST_BYTES)?;
    sealed.manifest_id = format!("{MANIFEST_PREFIX}{digest}");
    sealed.manifest_sha256 = digest;
    wire::canonical_with_max(&sealed, MAX_MANIFEST_BYTES)?;
    shape(&sealed, false)?;
    Ok(sealed)
}

/// Derives and seals the only complete manifest for a sealed ADR-0090 closure.
///
/// # Errors
///
/// Returns an error when the dependency is invalid or its projection exceeds the ceiling.
pub fn derive_structural_replay_manifest(
    decision_closure: &KernelDecisionReferenceClosure,
) -> Result<StructuralReplayManifest, DecisionCapsuleContractError> {
    let closure = primitives::reseal_decision_closure(decision_closure)?;
    derive_with_closure(&closure)
}

/// Decodes canonical manifest bytes and validates them against a sealed ADR-0090 closure.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or dependency-inconsistent bytes.
pub fn decode_structural_replay_manifest(
    bytes: &[u8],
    decision_closure: &KernelDecisionReferenceClosure,
) -> Result<StructuralReplayManifest, DecisionCapsuleContractError> {
    let value = wire::decode_typed(bytes, MAX_MANIFEST_BYTES)?;
    validate_structural_replay_manifest(&value, decision_closure)?;
    Ok(value)
}
